types/vote_set.go

```diff
@@ -128,20 +128,29 @@ func (voteSet *VoteSet) addVote(val *Validator, valIndex int, vote *Vote) (bool,
 		return false, 0, ErrVoteUnexpectedStep
 	}
 
+	var doubleSign bool
 	// If vote already exists, return false.
 	if existingVote := voteSet.votes[valIndex]; existingVote != nil {
 		if bytes.Equal(existingVote.BlockHash, vote.BlockHash) {
 			return false, valIndex, nil
 		} else {
-			// Check signature.
-			if !val.PubKey.VerifyBytes(SignBytes(config.GetString("chain_id"), vote), vote.Signature) {
-				// Bad signature.
-				return false, 0, ErrVoteInvalidSignature
-			}
-			return false, valIndex, &ErrVoteConflictingSignature{
-				VoteA: existingVote,
-				VoteB: vote,
+			doubleSign = true
+			log.Warn("Double sign detected!", "height", vote.Height, "round", vote.Round, "type", vote.Type, "index", valIndex)
+
+			// if its a preovte, or we already have a majority, return the double sign error.
+			// else, we may need to replace the vote if its part of a commit
+			// so we let it fall through
+			if vote.Type == VoteTypePrevote || voteSet.maj23Exists {
+				return false, valIndex, &ErrVoteConflictingSignature{
+					VoteA: existingVote,
+					VoteB: vote,
+				}
 			}
+
+			// subtract voting power from the existing vote
```

Comment:
the fix is really that peers will send random votes if recving peer's vote bit array is full. so this is just the accounting to subtract the old vote and apply the new one if/when its received. obviously not the most optimal approach, but eventually we see the right votes.